### PR TITLE
[REVIEW] Update get_ipc_handle() to use cuda driver API

### PR DESCRIPTION
**Pull request by shwina**
+ removes more legacy RMM bindings

### CODE REVIEW

**Comment by gmarkall** [comment]
At the moment this doesn't pass the Numba IPC tests, but it turns out not to have been passing before your PR - these tests were augmented since the EMM plugin was merged to RMM, but I didn't notice the failures start to occur because I don't regularly run the Numba testsuite with RMM. Exceptions are similar to:

```
======================================================================
ERROR: test_ipc_array (numba.cuda.tests.cudapy.test_ipc.TestIpcMemory) (index=None, foreign=True)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gmarkall/numbadev/numba/numba/cuda/tests/cudapy/test_ipc.py", line 190, in test_ipc_array
    self.check_ipc_array(index, foreign)
  File "/home/gmarkall/numbadev/numba/numba/cuda/tests/cudapy/test_ipc.py", line 172, in check_ipc_array
    ipch = devarr.get_ipc_handle()
  File "/home/gmarkall/numbadev/numba/numba/cuda/cudadrv/devicearray.py", line 316, in get_ipc_handle
    ipch = devices.get_context().get_ipc_handle(self.gpu_data)
  File "/home/gmarkall/numbadev/numba/numba/cuda/cudadrv/driver.py", line 1099, in get_ipc_handle
    return self.memory_manager.get_ipc_handle(memory)
  File "/home/gmarkall/numbadev/rmm/python/rmm/rmm.py", line 199, in get_ipc_handle
    ctypes.byref(ipchandle), memory.owner.handle,
AttributeError: 'ForeignArray' object has no attribute 'handle'
```

This is a similar issue to https://github.com/numba/numba/issues/5429 which is now fixed in Numba.

A working implementation requires the following change on top of this PR:

```diff
diff --git a/python/rmm/rmm.py b/python/rmm/rmm.py
index 23400e6..365de22 100644
--- a/python/rmm/rmm.py
+++ b/python/rmm/rmm.py
@@ -194,14 +194,13 @@ class RMMNumbaManager(HostOnlyCUDAMemoryManager):
         Get an IPC handle for the MemoryPointer memory with offset modified by
         the RMM memory pool.
         """
+        start, end = cuda.cudadrv.driver.device_extents(memory)
         ipchandle = (ctypes.c_byte * 64)()  # IPC handle is 64 bytes
         cuda.cudadrv.driver.driver.cuIpcGetMemHandle(
-            ctypes.byref(ipchandle), memory.owner.handle,
+            ctypes.byref(ipchandle), start,
         )
         source_info = cuda.current_context().device.get_device_identity()
-        ptr = memory.device_ctypes_pointer.value
-        start, end = cuda.cudadrv.driver.device_extents(memory)
-        offset = ptr - start
+        offset = memory.handle.value - start
         return IpcHandle(
             memory, ipchandle, memory.size, source_info, offset=offset
         )
```

This then makes the `get_ipc_handle()` implementation in RMM functionally identical to the one internal to Numba.

> It'd be a bit weird to test this functionality in `rmm`, since `get_ipc_handle` isn't exercising RMM in any way.

If we were to leave the implementation of `get_ipc_handle` in RMM, then I think it would make sense to have a test of it in RMM, because `RMMNumbaManager.get_ipc_handle` is part of RMM. However...

> @gmarkall @kkraus14 any comments here on why this implementation of `get_ipc_handle` can't/shouldn't live in numba, for example in the base class `HostOnlyCUDAMemoryManager`?

There is no reason we can't provide the `get_ipc_handle` implementation in Numba now that the implementations in Numba need to be identical. It shouldn't go in `HostOnlyCUDAMemoryManager` because that class is only for host allocations, and another EMM plugin might want to subclass it but not end up with a default implementation of `get_ipc_handle`. Instead I'd suggest a mixin that provides a default implementation of `get_ipc_handle`, that can be used in conjunction with `HostOnlyCUDAMemoryManager` if desired.

To summarise, I think the steps to take are:

1. Apply the above patch in this PR to make RMM work correctly with the Numba IPC tests again.
2. I'll make a PR to add a class that provides the implementation of `get_ipc_handle`.
3. Once 2 is done and available in a released version of Numba (hopefully 0.51), we can remove the implementation of `get_ipc_handle` from RMM and use the Numba-provided one.

**Comment by gmarkall** [comment]
Just realised I forgot to explain the rationale behind the suggested change - the problem with the code as-is is that the passed in `MemoryPointer`, `memory`, may have an owner that is not itself a `MemoryPointer` - this can happen when the owner is not a Numba device array, but instead is another library's array via the CUDA Array Interface. In this case, `memory.owner` has no `handle` attribute, as this is specific to the `MemoryPointer` in Numba. The suggested change makes `get_ipc_handle` work for Numba arrays and other arrays via the CUDA Array Interface by using the start address from `device_extents` instead of using `memory.owner.handle` - `memory.owner.handle` is expected to point to the start of an allocation for Numba device arrays.

**Comment by gmarkall** [comment]
> 2. I'll make a PR to add a class that provides the implementation of `get_ipc_handle`.

PR to add the implementation to Numba for RMM and other EMM plugins to use: https://github.com/numba/numba/pull/5851

**Comment by shwina** [comment]
Thanks @gmarkall! :) Applied your patch
